go/chat/boxer.go

```diff
@@ -269,49 +254,57 @@ func (b *Boxer) unboxMessageWithKey(ctx context.Context, msg chat1.MessageBoxed,
 			OutboxID:     hp.OutboxID,
 		}
 	default:
-		return unboxMessageWithKeyRes{},
+		return nil,
 			NewPermanentUnboxingError(NewHeaderVersionError(headerVersion,
 				b.headerUnsupported(ctx, headerVersion, header)))
 	}
 
-	if skipBodyVerification {
-		// body was deleted, so return empty body that matches header version
-		switch headerVersion {
-		case chat1.HeaderPlaintextVersion_V1:
-			return unboxMessageWithKeyRes{
-				messagePlaintext:      chat1.MessagePlaintext{ClientHeader: clientHeader},
-				headerHash:            headerHash,
-				headerSignature:       headerSignature,
-				senderDeviceRevokedAt: validity.senderDeviceRevokedAt,
-			}, nil
+	// Check for sender match on the inner and outer header.
+	if !clientHeader.Sender.Eq(boxed.ClientHeader.Sender) {
+		return nil, NewPermanentUnboxingError(fmt.Errorf("sender does not match"))
+	}
+	if !bytes.Equal(clientHeader.SenderDevice.Bytes(), boxed.ClientHeader.SenderDevice.Bytes()) {
+		return nil, NewPermanentUnboxingError(fmt.Errorf("sender device does not match"))
+	}
+
+	senderUsername, senderDeviceName, senderDeviceType := b.getSenderInfoLocal(
+		ctx, clientHeader.Sender, clientHeader.SenderDevice)
+
+	// create a chat1.MessageBody from versioned chat1.BodyPlaintext
+	// Will remain empty if the body was deleted.
+	var body chat1.MessageBody
+	if !skipBodyVerification {
+		bodyVersion, err := bodyVersioned.Version()
+		if err != nil {
+			return nil, NewPermanentUnboxingError(err)
+		}
+		switch bodyVersion {
+		case chat1.BodyPlaintextVersion_V1:
+			body = bodyVersioned.V1().MessageBody
 		default:
-			return unboxMessageWithKeyRes{},
-				NewPermanentUnboxingError(NewHeaderVersionError(headerVersion,
-					b.headerUnsupported(ctx, headerVersion, header)))
+			return nil,
+				NewPermanentUnboxingError(NewBodyVersionError(bodyVersion,
+					b.bodyUnsupported(ctx, bodyVersion, bodyVersioned)))
 		}
 	}
 
-	// create an unboxed message from versioned BodyPlaintext and clientHeader
-	bodyVersion, err := body.Version()
-	if err != nil {
-		return unboxMessageWithKeyRes{}, NewPermanentUnboxingError(err)
-	}
-	switch bodyVersion {
-	case chat1.BodyPlaintextVersion_V1:
-		return unboxMessageWithKeyRes{
-			messagePlaintext: chat1.MessagePlaintext{
-				ClientHeader: clientHeader,
-				MessageBody:  body.V1().MessageBody,
-			},
-			headerHash:            headerHash,
-			headerSignature:       headerSignature,
-			senderDeviceRevokedAt: validity.senderDeviceRevokedAt,
-		}, nil
-	default:
-		return unboxMessageWithKeyRes{},
-			NewPermanentUnboxingError(NewBodyVersionError(bodyVersion,
-				b.bodyUnsupported(ctx, bodyVersion, body)))
-	}
+	// create an unboxed message
+	return &chat1.MessageUnboxedValid{
+		ClientHeader:          clientHeader,
+		ServerHeader:          *boxed.ServerHeader,
+		MessageBody:           body,
+		SenderUsername:        senderUsername,
```

Comment:
Maybe. I'd like to punt this by leaving it unchanged and making a note to myself to look into it later. Part of the complication is that we may not have an easy way to get revoked device names.